### ISSUE TITLE
Local Developer Experience - Run multiple dapr apps using `dapr run -f`

### ISSUE THREAD

**Issue by shubham1172** [dapr/cli#1123]
## Describe proposal

This is related to the implementation of the proposal https://github.com/dapr/proposals/blob/main/0003-I-enhance-dapr-run-multiple-apps.md

### Targets for v1.10
The main requirements for the command:
- being able to configure multiple dapr apps from a single configuration file
- users should be able to use normal dapr CLI commands for self hosted mode against any apps that are started through `dapr run -f`

- Initial implementation will only support Linux & macOs.
- `dapr run -f <path>` will be an interactive process running completely for the lifecycle of applications, on exiting the process, 
- all other spawned processes will also quit.
- logs will be written to a predefined location. Users will need to manually tail the file. 

- [x] Get Sign off on feature proposal dapr/proposals#6
- [x] https://github.com/dapr/cli/issues/1141
- [x] https://github.com/dapr/cli/issues/1142
- [x] https://github.com/dapr/cli/issues/1143
- [x] Docs for `dapr run -f`
- [x] https://github.com/dapr/cli/issues/1153

Future enhancements
- [ ] https://github.com/dapr/cli/issues/1220
- [ ] https://github.com/dapr/cli/issues/1221
- [ ] https://github.com/dapr/cli/issues/1228
- [ ] https://github.com/dapr/cli/issues/1204
- [ ] https://github.com/dapr/cli/issues/1144
- [ ] https://github.com/dapr/cli/issues/1145
- [ ] https://github.com/dapr/cli/issues/1146
- [ ] https://github.com/dapr/cli/issues/1147
- [ ] https://github.com/dapr/cli/issues/1154
- [ ] https://github.com/dapr/cli/issues/1155

## ~~Describe the proposal (Old version)~~

This issue derives from https://github.com/dapr/community/issues/207 to propose two new commands - 
`dapr compose up` and `dapr compose down`.

### Compose file structure

```yaml
version: 1
common:
  environment:  # any environment variable shared among apps
    - DEBUG: true
services:
  - id: kafka
    image: docker.io/bitnami/kafka
    port: 4040
    environment:
      - KAFKA_BROKER_ID: 1
apps:
  - id: webapp
    root_directory: ./webapp/
    components_directory: ./webapp/components # can be default by convention too, ignore if dir is not found.
    config_file: ./webapp/config.yaml # can be default by convention too, ignore if file is not found.
    protocol: HTTP
    port: 8080
    command: python3 app.py
  - id: backend
    root_directory: ./backend/
    components_directory: ./backend/components # can be default by convention too, ignore if dir is not found.
    config_file: ./backend/config.yaml # can be default by convention too, ignore if file is not found.
    protocol: GRPC
    port: 3000
    env:
      - DEBUG: false
    command: ./backend
```

### User experience

- The user can run `dapr compose up` to start these services and apps
- If something crashes, it will be restarted with exponential back-off until a back-off limit
- If ports are missing, compose automatically finds a free port
- The user can run `dapr compose down` to gracefully stop everything

### Open questions

1. Should we separate `services` into another proposal?
1. Should we support running apps in containers directly?
1. Should we support running alternate container runtime (like podman) since CLI already does that?

## Release Note

RELEASE NOTE: **ADD** Dapr compose up and down

**Comment by mukundansundar** [comment]
Related PR dapr/proposals#6

**Comment by mukundansundar** [comment]
Issues to be created 
- Implement YAML template processing with UTs for the run configuration file. (Take call on whether multiple Yaml defs in single file is going to be processed or not. (Initial thoughts do not allow.))



- Add capability to look for resources under given input directory and fallback to default if not present with UTs (folder org structure) (See if it can be implemented for normal dapr run without breaking applications) (preview flag?)



- Implement basic dapr run -f with E2Es (run app commands and daprd command, if app command is not given run only daprd. Generate the ports automatically and inject as env var into app (HTTP, GRPC, Internal GRPC and Metrics). Write logs to file and STDOUT See `io.MultiWriter`)(Run apps sequentially)



- (Optional if not in future release)Enhance the template to have a `depends_on` field and run apps in parallel




- Enhance dapr run to store logs temporarily for any run (for future releases. `~/.dapr/app_id/logs/`, or if -f `<app-folder>/.dapr/logs`. )



- Enhance dapr logs to view logs (Use the folder path defined above and tail log files)



- Add a detached mode for run command

**Comment by mukundansundar** [comment]
@yaron2 Please take a look on the issues list ... Will be creating them soon

**Comment by mukundansundar** [comment]
> Related PR [dapr/proposals#6](https://github.com/dapr/proposals/pull/6)

This proposal has been accepted for implementation ...

**Comment by philliphoff** [comment]
@mukundansundar I've created several proposals (#1153, #1154, #1155) based on our earlier conversation as to how best accommodate tooling with this new feature.

**Comment by greenie-msft** [comment]
@mukundansundar is there a docs issue for this effort already?

**Comment by hhunter-ms** [comment]
> @mukundansundar is there a docs issue for this effort already?

Docs issue: dapr/docs#3047

**Comment by mukundansundar** [comment]
moving this Epic to another milestone ... initial implementation as part of 1.10 is done as seen in the linked issues above.

**Comment by philliphoff** [comment]
Where does "native" Windows support sit relative to other tasks (if at all)?

**Comment by artursouza** [comment]
@mukundansundar Please, create a new issue for enhancements in 1.10
